platform/ScopedLock.h

```diff
@@ -0,0 +1,78 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_SCOPEDLOCK_H
+#define MBED_SCOPEDLOCK_H
+
+#include "platform/NonCopyable.h"
+
+namespace mbed {
+
+/** \addtogroup platform */
+/** @{*/
+/**
+ * \defgroup platform_ScopedLock ScopedLock functions
+ * @{
+ */
+
+/** ScopedLock class, RAII-style mechanism for owning a lock of Lockable object for the duration of a scoped block
+ *
+ * @note To be used with types implementing BasicLockable concept
+ *
+ * Usage:
+ * @code
+ *
+ * template<typename Lockable>
+ * void foo(Lockable& lockable) {
```

Comment:
Maybe it would be a good thing to use a concrete type like a `Mutex` for the first example and be more generic with the second example.

Reply:
done